Add significantly more debug logging to UIAHandler

michaelDCurran opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 

Please initially open PRs as a draft. See https://github.com/nvaccess/nvda/wiki/Contributing
-->

### Link to issue number:
None.

### Summary of the issue:
Although NVDA has a specific uIA debug logging category, this does not really enable all that many more log messages a part from some debug warnings. thus it is almost useless to ask for a user's log with this flag enabled.

 
### Description of user facing changes
Setting NVDA's log level to debug, and enabling the UIA debug logging category in NVDA's advanced settings panel, will now produce significantly more logging for UIA. This includes logging when all UIA event handler methods are called, along with their parameters, and logging for utility utility methods such as checking if a window or element is native UIA.
 
### Description of development approach
Added a bunch of methods to UIAHandler for generating a string representation for debugging, for things such as UIA elements, property IDs, event IDs, notification types etc.
Added conditional debug logging to all UIA event handler methods, plus isUIAWindow, isNativeUIAElement and getNearestWindowhandle.
 
### Testing strategy:
Ran NVDA for a day or so, performing general tasks, using multiple apps that are native UIA.
Used the new debug logging to aide in investigating QT issues.

### Known issues with pull request:
None known.

### Change log entries:
New features
Changes
Bug fixes
For Developers
* the UIA debug logging category when enabled now produces significantly more logging for UIA event handlers and utilities.
### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

michaelDCurran committed:
UIAHandler: move base cacheRequest propertyIDs into their own set at global level.

michaelDCurran committed:
UIAHandler: add much greater debug logging to all UIA event handlers.

josephsl commented:
Subscribing as it will impact Event Tracker add-on once this PR is merged (if merged, I will declare Event Tracker as a legacy add-on). Thanks.

seanbudd commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -1292,6 +1302,11 @@ def __init__(self,windowHandle=None,UIAElement=None,initialUIACachedPropertyIDs=
 		if not windowHandle:
+			if UIAHandler._isDebug():
+				log.debug(
+					f"No windowHandle for UIA NvDAObject. "
```
Fix capitalization of NVDA

```suggestion
					f"No windowHandle for UIA NVDAObject. "
```

seanbudd replied:
this is still an issue, same with the unnecessary f-strings.
However, as the remaining changes are trivial, I am approving in the mean-time

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -49,6 +49,14 @@
 
 
+baseCachePropertyIDs = set([
+	UIA.UIA_FrameworkIdPropertyId, UIA.UIA_AutomationIdPropertyId, UIA.UIA_ClassNamePropertyId,
```
Could these be separate to 1 per new line?
It improves visual readability and makes diffs simpler if items are added/removed.

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -49,6 +49,14 @@
 from . import remote as UIARemote
 
 
+baseCachePropertyIDs = set([
```
```suggestion
baseCachePropertyIDs = {
```

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
 		UIA.IUIAutomationActiveTextPositionChangedEventHandler,
 	]
 
+	#: A cache of UIA notification kinds  to friendly names for logging
```
Fix double space

```suggestion
	#: A cache of UIA notification kinds to friendly names for logging
```

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
+		"""
+		name = self._notificationKindsToNamesCache.get(notificationKind)
+		if not name:
+			name = "unknown notification kind {notificationKind}"
```
should this be an f-string? I cant't find a corresponding `.format`

```suggestion
			name = f"unknown notification kind {notificationKind}"
```

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
+		Generates a string representation of the given UIA notification kind,
+		suitable for logging.
+		This is the name part of the NotificationKind_* constant.
+		If a matching  constant can not be found,
```
Fix double space

```suggestion
		If a matching constant can not be found,
```

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
+		Generates a string representation of the given UIA notification processing value,
+		suitable for logging.
+		This is the name part of the NotificationProcessing_* constant.
+		If a matching  constant can not be found,
```
double space

```suggestion
		If a matching constant can not be found,
```

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
+		"""
+		name = self._notificationProcessingValuesToNamesCache.get(notificationProcessing)
+		if not name:
+			name = "unknown notification processing value {notificationProcessing}"
```
same here with f-string

michaelDCurran replied:
```suggestion
			name = f"unknown notification processing value {notificationProcessing}"
```

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
+		"""
+		name = self._eventIDsToNamesCache.get(eventID)
+		if not name:
+			name = "unknown event ID {eventID}"
```
same here with f-string

michaelDCurran replied:
```suggestion
			name = f"unknown event ID {eventID}"
```

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
+		Generates a string representation of the given UIA event ID,
+		suitable for logging.
+		This is the name part of the UIA_*EventId constant.
+		If a matching  constant can not be found, then a string representation of the ID itself is used.
```
Double space

```suggestion
		If a matching constant can not be found, then a string representation of the ID itself is used.
```

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -572,9 +702,8 @@ def IUIAutomationEventHandler_HandleAutomationEvent(self,sender,eventID):
-						"HandleAutomationEvent: Dropping textChange event from element "
-						f"with class name {repr(sender.currentClassName)} "
-						f"and automation ID {repr(sender.CachedAutomationID)}"
+						f"HandleAutomationEvent: Dropping textChange event "
```
Remove unnecessary f string

```suggestion
						"HandleAutomationEvent: Dropping textChange event "
```

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -612,26 +743,42 @@ def IUIAutomationEventHandler_HandleAutomationEvent(self,sender,eventID):
 
 	def IUIAutomationFocusChangedEventHandler_HandleFocusChangedEvent(self,sender):
+		if _isDebug():
+			log.debug(f"handleFocusChangedEvent called with  element {self.getUIAElementDebugString(sender)}")
```
Fix double space

```suggestion
			log.debug(f"handleFocusChangedEvent called with element {self.getUIAElementDebugString(sender)}")
```

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -722,6 +896,8 @@ def IUIAutomationPropertyChangedEventHandler_HandlePropertyChangedEvent(self,sen
 			and self.clientObject.compareElements(focus.UIAElement, sender)
 		):
+			if _isDebug():
+				log.debug(f"propertyChange event is for  focus")
```
Fix double space

```suggestion
				log.debug(f"propertyChange event is for focus")
```

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -790,13 +986,22 @@ def IUIAutomationNotificationEventHandler_HandleNotificationEvent(
 			return
+		if _isDebug():
+			log.debug(
+				f"Queuing  UIA_notification NVDA event "
```
remove f string, fix double space

I'm not sure if this is an issue, but there are many strings that are unnecessarily marked as an f-string (i.e. no special formatting). 
I have seen reviews in the past encouraging f-strings only when required.

```suggestion
				"Queuing UIA_notification NVDA event "
```

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -904,23 +1141,49 @@ def _isUIAWindowHelper(self,hwnd):
 		return bool(res)
 
-	def isUIAWindow(self,hwnd):
+	def isUIAWindow(self, hwnd: int, isDebug=False) -> bool:
```
Add extra type information.

```suggestion
	def isUIAWindow(self, hwnd: int, isDebug: bool = False) -> bool:
```

seanbudd commented:
Mostly trivial changes remaining

michaelDCurran committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <sean@nvaccess.org>

michaelDCurran committed:
UIAHandler: linting

michaelDCurran committed:
Merge branch 'master' into UIADebugging

michaelDCurran committed:
UIAHandler: any f-strings that don't contain any templating are now standard strings.

michaelDCurran committed:
UIAHandler: make sure to pass isDebug argument to _isUIAWindowHelper so debug messages are produced if isUIAWindow is also being debugged.

michaelDCurran commented:
@seanbudd  I've changed all  unnecessary f-strings to regular strings, and also added a missing  isDebug argument to the call to _isUIAWindowHelper. So another review please.

michaelDCurran committed:
Update what's new
